Implement a table-like shortcut to rpm macros in Lua

pmatilai opened the pull request:
Add rpm macro context as a global table-like entity named 'm' for
typing convenience into the Lua environment.

Support basic access with table indexing syntaxes (m[k] and m.k),
undefined macros return nil here. As a specialty, parametric macros
are returned as native callable variadic Lua functions, arguments
are converted to strings.

pmatilai commented:
See the docs + test cases in the commit for details, if something isn't covered then do ask.

There's ample potential for bikeshedding here (eg what to call the global table), should it do some sort of auto-quoting for empty strings as arguments etc.

Tagging @hroncok @nim-nim @mlschroe @jasontibbits for past interest/activities in this area, feel free to invite more.

nim-nim commented:
If that works that will be terrific! However as usual the devil is in the detail, is it available as a pre-built rpm somewhere or do one need to build his own binary?

pmatilai commented:
https://copr.fedorainfracloud.org/coprs/pmatilai/rpm-snapshot/ has pre-built binaries for Fedora, updated manually once/twice a week. Usually those are strictly from rpm git master but it's not like I can't make exceptions... https://copr.fedorainfracloud.org/coprs/pmatilai/rpm-snapshot/build/1705701/ has this PR included (this also includes the native option/argument stuff from last week of course)

hroncok commented:
> There's ample potential for bikeshedding here (eg what to call the global table)

Indeed :) `m` is nice and short, but possibly a bit too much? What about including it in the `rpm` namespace as `rpm.m`?

pmatilai commented:
It stops being nice and short if we put it into the rpm namespace :-/ 

Unlike a generic Lua library extension, rpm owns the environment in which these things run so I don't see a reason we couldn't maximise the convenience. Too bad `%` is not a legitimate character in Lua names... OTOH, `_` is and could be used, but underscores tend to hint at something you shouldn't touch, which is not the case here.

voxik commented:
> > There's ample potential for bikeshedding here (eg what to call the global table)
> 
> Indeed :) `m` is nice and short, but possibly a bit too much? What about including it in the `rpm` namespace as `rpm.m`?

I agree with @hroncok. I think the code should be readable on the first place. `m` does not add to readability. Not mentioning if somebody wants to find reference using search engine.

pmatilai commented:
Could of course call the table `macro`to make it blindingly obvious what that is, although I do think the context makes it quite clear anyhow. It's not like rpm macros ever were nice or readable to begin with :stuck_out_tongue_winking_eye: 

Macros are by far the most accessed thing from rpm Lua so I think sheer convenience should be high on priorities.

pmatilai commented:
Other random remarks of the day:
- We could of course put it into rpm namespace, and macro-heavy users can declare their own shortcut, eg `m = rpm.macro; m._libdir = "foo"`, but then if every Lua macro starts with that, we've failed somewhere
- This doesn't enable *anything* new, it's all just for convenience. And as such, all this could be also implemented in pure Lua instead. Doing that might open some other possibilities, OTOH keeping things in one place and one language has other benefits.
- I realized this is missing the mapping for Lua native deletion, ie `m[key] = nil` to undefine. Will add in a later version.

mlschroe commented:
You can easily implement quoting by always surrounding the arguments with 0x1f characters, see splitQuoted() and the %quote macro.

pmatilai commented:
Yup. The reason it's not there now is that from wherever I had this idea it might prevent macro expansion in said arguments, and somehow never got around to check what it actually did :roll_eyes:  (I'm getting old, it's just three years since I added that thing...)

Now actually checking, yeah always quoting the args is quite clearly the right thing to do for this.

pmatilai commented:
Updated:
- rename the global `m` table to `macro`, it's indeed better for readability
- implement undefine via nil assignment
- automatically quote parametric macro arguments when called as Lua function
- rebased
- new build of this state at https://copr.fedorainfracloud.org/coprs/pmatilai/rpm-snapshot/build/1706906/

mlschroe commented:
Oh, you want the arguments in a macro call expanded? I was going to suggest that we should add code to make them stay unexpanded (i.e. double the '%' chars). There's merit in both sides, I guess.

pmatilai commented:
Calling a parametric macro as a function vs traditional rpm.expand() needs to behave identically as the *called* macro will have no idea about any of this. Auto-quoting is of course already on a slippery slope.

mlschroe commented:
No, this is not about the called macro. My point is with macro calls via rpm.expand, we expand macros first and then split into args. With your new call from lua we already provide the args, so it's not clear if we want expansion or not. Both works, it's just a matter of documentation.

pmatilai commented:
But in the function call, we just catenate the args into a string that gets passed plain old rpmExpand(), so if one method expands them and the other doesn't, you can get quite different results. Which doesn't seem sane.

nim-nim commented:
Looking back on my code, I see it required both the expanded and non expanded variants. The general case is expansion (recursively, till there is nothing else to process). However in less frequent but tricky cases expanding can be premature and you really need to pass things as is (I've come to think of those as pointers/references to stuff that can not be expanded till later when the data context is populated enough for expansion to succeed).

However in the non expanded case you still need to test for macro existence, so non expanded is not a plain %{?foo} but %{?foo} if foo exists and nil otherwise.

If there is an expanded vs non expanded variant I’d suggest making the expanded variant fully expanded and interpreted, ie re-expanding till there is nothing more to expand, stripping quoted lines, stripping empty lines, real booleans, multi-line values represented as a table of values.

The advantage of a fully expanded variant is that you kill all the macro heisenbugs where initial unit testing is fine but field use fails because packagers stack macros and stacking adds levels of expansion not tested against initially.

pmatilai commented:
This is starting to spin out of scope...

This PR is not about changing parametric macro semantics at all, it's about making them more accessible from Lua, and basically I think results need to be identical between `rpm.expand('%foo [...])` and `macro.foo([...])` or everybody will go crazy. Which is why I'm starting to think the automatic quoting is a bad idea, because it causes a deviation, and where there's one... it's easy to end up with more, whether due to side-effect pileup or otherwise.

We "just" changed the macro arguments to be expanded in 4.14 and it took multiple attempts to get the worst kinks out, I'm not all that eager to experiment in that area but if people want to discuss that, please open a separate ticket.

mlschroe commented:
But then you should change the code so that it does not accept more than one argument. Everything else is not "identical" and does not match the user expectation.

Btw, please do not use %quote, this will break if the argument contains an unmatched '}' character. It's much saner to simply append/prepend that 0x1f character even if it somewhat breaks abstraction.

mlschroe commented:
(Hmm, an unmatched '}' will break the expansion anyway, so nevermind about not using %quote.)

pmatilai commented:
> But then you should change the code so that it does not accept more than one argument. Everything else is not "identical" and does not match the user expectation.

Hmm. I fail to see the problem here, it's just a shortcut for building that string yourself, ie these two should be equal:

```
res = rpm.expand('%foo'..' '..opt..' '..arg1..' '..arg2)
res = macro.foo(apt, arg1, arg2)
```

We could of course have the latter understand other forms too, such as passing (opt, arg) tables (which again are just automatically flattened to the actual macro string).

voxik commented:
>     * rename the global `m` table to `macro`, it's indeed better for readability

Thx, just nitpicking, shouldn't it be `macros`? The `macro._libdir == '/usr/lib64'` is probably Ok, but `macro['zzz']` is questionable.

pmatilai commented:
The plural form could be argued, yes. I considered it, but somehow it seems weird to me in this context, in particular the dot-syntax (eg `macros.with('foo')`). Do you have some concrete reason why `macro['_libdir']` seems objectionable to you? I'm just curious, as I actually find that more natural than the plural form, but I'd be hard-pressed to explain *why* :sweat_smile:

pmatilai commented:
Updated:
- macro call arguments are auto-quoted if passed as a table
- more docs on the behavior
- build the strings on Lua side and use Lua rpm.expand() call consistently for all macro expansions

voxik commented:
Typically if I have some `macro`, then the collection of them is plural `macros`. If you want to iterate over the collection, then you want to have something like `for macro in macros do ...`. Of course you can have `for m in macro do ...` or something similar, but that does not help with readability.

Of course it is questionable, if iteration over macros it the thing.

pmatilai commented:
Okay, iteration is a fair point. The catch is that we don't support that at all, and I'm not sure we ever will.

pmatilai commented:
After some more consideration, renamed to `macros`. This is more in line with our other globals (sources, patches) and leaves room for iteration options in the future.

pmatilai commented:
Rebased again to include the other new Lua/macro stuff and a new snapshot build at https://copr.fedorainfracloud.org/coprs/pmatilai/rpm-snapshot/build/1712012/

mlschroe commented:
Regarding the macro call feature: I was going to suggest something like that. Perl, for example, offers both `system(STRING)` and `system(LIST)`. The STRING version uses the shell to do the call, so you get argument splitting and variable expansion. The LIST version does a verbatim call without any expansion.

Now I just need to convince you that:
- you should drop that misleading autojoin feature and just accept one optional argument which can either be a string or a table
- you should add a `rpmExpandThisMacro(rpmMacroContext mc, const char *n, ARGV_t args)` function which can be used to expand macro `n` with the provided args without needing the rpmExpand arg hackery. Note that this call would not expand the arguments.

Btw, what about builtins? They currently cannot be accessed.

hroncok commented:
Indeed, either 1 string or 1 list (table) sounds more natural to me.

pmatilai commented:
Now that the table version exists, I don't feel any particular need to hang on to the multi-argument auto-join thing. The convenience of not having to manually and tediously construct the string was the point of that, but since *that* can be avoided with the table syntax, eliminating the string catenation neatly eliminates the problem of quoting - it's simply up to the caller. I'll update that part.

As for builtins, I know. One possibility would be hiding the different calling convention in the call, so eg `rpm.expand("%{uncompress:...}")` becomes just `macros.uncompress(...)`, but I actually think it'd be saner to expose them via an entirely different "namespace" because builtins can't be overridden, so they'd need annoying extra checks in the macros table. Maybe `builtin.uncompress(...)` or something.

The macro argument expansion is a stickier issue because it's a rather fundamental part of the parametric macro calling convention. If different means of calling the same macro can behave differently, it can create strange situations (and bugs). I think. Note that I've been toying with an idea of some sort of "native Lua macro" concept that would by definition bypass the normal %{1} etc argument processing and just call the other macro as a normal Lua function, but that's just a very vague idea at this point.

pmatilai commented:
On a related note on the builtins: I've been thinking about turning the builtins into parametric macros, or at least blur the difference considerably (and allow dynamic registering of builtins). The syntax differs a bit and the calling convention much more, so it's not entirely straightforward/obvious how to proceed there. If they're technically merged into regular macros, the macros/builtins split wouldn't probably make much sense.

pmatilai commented:
Updated to have function calls only accept a single string or table.

pmatilai commented:
>  you should add a `rpmExpandThisMacro(rpmMacroContext mc, const char *n, ARGV_t args)` function which can be used to expand macro `n` with the provided args without needing the rpmExpand arg hackery. Note that this call would not expand the arguments.

@mlschroe , it's not so much that I disagree, it's just that it's easier said than done :sweat_smile: 
Took a good part of the afternoon but I have it more or less working now, I'll try to get it into a submittable state tomorrow or so.

pmatilai commented:
> I'll try to get it into a submittable state tomorrow or so.

...except that nope, didn't see a way to achieve this with a reasonable effort. Doesn't necessarily mean there isn't one, but I lost my appetite for now. I might revisit at some later time, but if somebody wants to pick up the pieces in the meanwhile, feel free.

mlschroe commented:
Sorry, I didn't want to discourage you. I'll open a pull request with that rpmExpandThisMacro() implementation.

pmatilai commented:
No worries. I got tangled in the argument/option parsing/passing recursion - I've a feeling the mb->args vs argument macros and all isn't quite right as it is, and just ran out of steam for now. So I was actually kinda hoping you'd say that...

Meanwhile, here's a related step towards making builtins more accessible: #1406
